# Preparing Input Data for Amazon Personalize

In this notebook, we'll work through **selecting and preparing** historical training data for Amazon Personalize - uploading the prepared files to Amazon S3.

> ⚠️ We assume you've *already* set up an S3 bucket and IAM execution role for Personalize to access the bucket (either in the AWS console or by running notebook [00_Environment_Setup.ipynb](00_Environment_Setup.ipynb)) and `%store`d this information as we did in the notebook.

## Introduction

Amazon Personalize provides 3 types of "recipe" (algorithm), for solving different recommendation tasks. Some categories include multiple recipes, as follows:

1. [**User Personalization**](https://docs.aws.amazon.com/personalize/latest/dg/user-personalization-recipes.html) - Recommend relevant items for a given user (with some business rule filtering, if required)
    - The [**User-Personalization**](https://docs.aws.amazon.com/personalize/latest/dg/native-recipe-new-item-USER_PERSONALIZATION.html) recipe is the most fully-featured and typically **recommended for all use-cases in this category**
    - The legacy **HRNN-...** recipes in this category should normally be substituted for User-Personalization
    - The [**Popularity-Count**](https://docs.aws.amazon.com/personalize/latest/dg/native-recipe-popularity.html) is a useful **baseline** recipe for contextualizing performance metrics: e.g. "What metrics would I get *just by always recommending the most popular items*)
2. [**Personalized Ranking**](https://docs.aws.amazon.com/personalize/latest/dg/personalized-ranking-recipes.html) - Prioritize relevant items **from a provided shortlist** for a given user (for supplying lists, rather than applying filter rules)
3. [**Related Items**](https://docs.aws.amazon.com/personalize/latest/dg/related-items-recipes.html) - Recommend relevant items for a given **item** in context (e.g. "customers also bought")

No matter the use case, the algorithms all share a base of learning on **user-item-interaction data** - a set of *events* defined by 3 core required attributes:

1. **User_ID** - The user who interacted
2. **Item_ID** - The item the user interacted with
3. **Timestamp** - The time at which the interaction occurred

Personalize also defines certain optional event attributes you can add:

4. **Event_Type** - Categorical label of an event (browse, purchased, rated, etc).
5. **Event_Value** - A value corresponding to the event type that occurred. As a best practice, we generally aim for normalized values between 0 and 1 to be comparable between event types. For example, if there are three phases to complete a transaction (clicked, added-to-cart, and purchased), then there would be an event_value for each phase as 0.33, 0.66, and 1.0 respectively.

...and users can also add **custom interaction attributes** which some recipes can include in the generated model: E.g. location, device type, etc.

In addition to this core **'Interactions' dataset**, we may *optionally* provide:

- **Users metadata** - a table mapping each `USER_ID` to additional custom attributes such as gender, age group, etc.
- **Items metadata** - a table mapping each `ITEM_ID` to additional custom attributes such as brand, category, etc.

More detail on the schema and requirements for Amazon Personalize training data is given in the [Datasets and Schemas](https://docs.aws.amazon.com/personalize/latest/dg/how-it-works-dataset-schema.html) and [Preparing and Importing Data](https://docs.aws.amazon.com/personalize/latest/dg/data-prep.html) sections of the Amazon Personalize Developer Guide.

## Choosing a Data Source

> ⚠️ **Remember:** Only the *Interactions* dataset is mandatory, and all recipes in Amazon Personalize at the time of writing use this as their primary learning source - with secondary input from other metadata.
>
> See the official [Amazon Personalize Cheat Sheet](https://github.com/aws-samples/amazon-personalize-samples/blob/master/PersonalizeCheatSheet2.0.md) for more guidance on **what use-cases Amazon Personalize is a good fit for**, and where you might consider alternative techniques instead.

Many use-cases *do* generate this kind of interaction history data - for example:

1. Video-on-demand applications
1. E-commerce platforms
1. Social media aggregators / platforms

...And we should be reasonably well-placed as long as our use case meets some basic criteria like:

- Authenticated users
- At least 50 unique users
- At least 100 unique items
- Several (and prefarably at least 2 dozen) interactions for each user

However, your historical data will typically not arrive in a perfect form & compatible schema for Personalize - so our first task will be to procure and **re-structure** the training data.

In this example, we'll use the [**MovieLens dataset**](https://grouplens.org/datasets/movielens/) to explore the service: A set of *movie reviews*.

## Fetching the Data (MovieLens Example)

The full MovieLens dataset includes over 25 million interactions (reviews) and a rich collection of metadata for items (movies).

There's also a smaller published extract of the dataset available, which we'll use by default here to shorten training times while still demonstrating the same capabilities. Set `USE_FULL_MOVIELENS` to `True` below if you'd like to use the full dataset instead:

In [1]:
USE_FULL_MOVIELENS = False

In [2]:
data_dir = "poc_data"
%store data_dir

!mkdir -p $data_dir

# Download and extract the dataset:
if not USE_FULL_MOVIELENS:
    !cd $data_dir && wget -N http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
    !cd $data_dir && unzip -o ml-latest-small.zip
    dataset_dir = data_dir + "/ml-latest-small/"
else:
    !cd $data_dir && wget -N http://files.grouplens.org/datasets/movielens/ml-25m.zip
    !cd $data_dir && unzip -o ml-25m.zip
    dataset_dir = data_dir + "/ml-25m/"
%store dataset_dir

Stored 'data_dir' (str)
--2021-03-23 06:38:54--  http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
Resolving files.grouplens.org (files.grouplens.org)... 128.101.65.152
Connecting to files.grouplens.org (files.grouplens.org)|128.101.65.152|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 978202 (955K) [application/zip]
Saving to: ‘ml-latest-small.zip’

ml-latest-small.zip 100%[===================>] 955.28K   821KB/s    in 1.2s    

2021-03-23 06:38:56 (821 KB/s) - ‘ml-latest-small.zip’ saved [978202/978202]

Archive:  ml-latest-small.zip
   creating: ml-latest-small/
  inflating: ml-latest-small/links.csv  
  inflating: ml-latest-small/tags.csv  
  inflating: ml-latest-small/ratings.csv  
  inflating: ml-latest-small/README.txt  
  inflating: ml-latest-small/movies.csv  
Stored 'dataset_dir' (str)


Take a look at the data files you've downloaded:

In [3]:
!ls $dataset_dir

links.csv  movies.csv  ratings.csv  README.txt	tags.csv


At present not much is known except that we have a few CSVs and a readme. Let's output the readme to learn more!

In [4]:
!pygmentize $dataset_dir/README.txt

Summary

This dataset (ml-latest-small) describes 5-star rating and free-text tagging activity from [MovieLens](http://movielens.org), a movie recommendation service. It contains 100836 ratings and 3683 tag applications across 9742 movies. These data were created by 610 users between March 29, 1996 and September 24, 2018. This dataset was generated on September 26, 2018.

Users were selected at random for inclusion. All selected users had rated at least 20 movies. No demographic information is included. Each user is represented by an id, and no other information is provided.

The data are contained in the files `links.csv`, `movies.csv`, `ratings.csv` and `tags.csv`. More details about the contents and use of all these files follows.

This is a *development* dataset. As such, it may change over time and is not an appropriate dataset for shared research results. See available *benchmark* datasets if that is your intent.

This and other GroupLens data sets are publicly available for down

From the README, we see there is a file `ratings.csv` that should form the basis of our **interactions** data - after all, rating a film definitely is a form of interacting with it!

The dataset also has some genre information as some movie genome data, which looks like a good source of **item metadata**.

As a publicly released dataset, there's not really any useful **user metadata** available here for us to model with - so we'll largely ignore this feature of Personalize in our example... But you can take the item metadata process as a good guide for your own experiments.

## Preparing Interactions Data

To get started exploring our data, we'll first import a few useful libraries:

In [5]:
# Python Built-Ins:
from datetime import datetime

# External Dependencies:
import boto3  # AWS SDK for Python
import pandas as pd  # DataFrame (table) manipulation tools

First we'd like to load our data and check the format, scope, and any gaps - to understand what might need adapting for Amazon Personalize:

In [6]:
original_data = pd.read_csv(dataset_dir + "/ratings.csv")

print(original_data.info())
original_data.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB
None


userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

From the above, you can see that there are a total of (25,000,095 for full 100836 for small) entries in the dataset, with 4 columns, and each cell has been interpreted as int64 format, except for the rating which has been loaded as float64.

We see that there are no null entries in any column (non-null counts all match), but may also be good to check the extent & basic statistics of the columns in case there's anything unexpected there:

In [7]:
original_data.describe()

userId        movieId         rating     timestamp
count  100836.000000  100836.000000  100836.000000  1.008360e+05
mean      326.127564   19435.295718       3.501557  1.205946e+09
std       182.618491   35530.987199       1.042529  2.162610e+08
min         1.000000       1.000000       0.500000  8.281246e+08
25%       177.000000    1199.000000       3.000000  1.019124e+09
50%       325.000000    2991.000000       3.500000  1.186087e+09
75%       477.000000    8122.000000       4.000000  1.435994e+09
max       610.000000  193609.000000       5.000000  1.537799e+09

The Ranges seem to be generally healthy for our `userId`, `movieId` and `rating` columns - and the inferred int64 & float64 formats clearly seem suitable for these fields.

However, we need to dive deeper to understand the timestamps in the data. Amazon Personalize [requires](https://docs.aws.amazon.com/personalize/latest/dg/data-prep-formatting.html#timestamp-data) timestamps in [Unix Epoch](https://en.wikipedia.org/wiki/Unix_time) format.

Currently, the timestamp values are not human-readable. So let's grab an arbitrary timestamp value and check whether it seems consistent.

In [8]:
arb_time_stamp = original_data.iloc[50]["timestamp"]
print(arb_time_stamp)
print(datetime.utcfromtimestamp(arb_time_stamp).strftime("%Y-%m-%d %H:%M:%S"))

964982681.0
2000-07-30 18:44:41


Great! This date seems to be within expected range, so we can continue formatting the rest of the data.

The correspondence is pretty clear from `userId` to `USER_ID`, and from `movieId` to `ITEM_ID`... But what about `rating`?

We can define a single `EVENT_TYPE` of "review", and use rating as our `EVENT_VALUE` field: Which will allow us to **threshold filter** our dataset to build models that consider only higher-rating events (e.g. to build a model more likely to recommend movies the user will actually *like*, not just what they're likely to review).

...So the only data preparation we'll need to do in this example is to rename our columns:

In [9]:
interactions_df = original_data.copy()
interactions_df["EVENT_TYPE"] = "review"
interactions_df.rename(
    columns={
        "userId": "USER_ID",
        "movieId": "ITEM_ID",
        "rating": "EVENT_VALUE",
        "timestamp": "TIMESTAMP",
    },
    inplace=True,
)
print(interactions_df.info())
interactions_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   USER_ID      100836 non-null  int64  
 1   ITEM_ID      100836 non-null  int64  
 2   EVENT_VALUE  100836 non-null  float64
 3   TIMESTAMP    100836 non-null  int64  
 4   EVENT_TYPE   100836 non-null  object 
dtypes: float64(1), int64(3), object(1)
memory usage: 3.8+ MB
None


USER_ID  ITEM_ID  EVENT_VALUE  TIMESTAMP EVENT_TYPE
0        1        1          4.0  964982703     review
1        1        3          4.0  964981247     review
2        1        6          4.0  964982224     review
3        1       47          5.0  964983815     review
4        1       50          5.0  964982931     review

To import this dataset to Amazon Personalize, we'll need to save it in **CSV format** on **Amazon S3**, so first let's save the file to a CSV here on our notebook:

In [10]:
interactions_filename = "interactions.csv"
interactions_path = f"{data_dir}/{interactions_filename}"
%store interactions_path

interactions_df.to_csv(interactions_path, index=False)

Stored 'interactions_path' (str)


...And then upload this file to the bucket we prepared in [00_Environment_Setup.ipynb](00_Environment_Setup.ipynb).

In [11]:
# Replace the below line with e.g. region = "ap-southeast-1" if you didn't run notebook 0
%store -r region
assert isinstance(region, str), "`region` must be a region name string e.g. 'us-east-1'"

# Replace the below line with e.g. bucket_name = "DOC-EXAMPLE-BUCKET" if you didn't run notebook 0
%store -r bucket_name
assert isinstance(bucket_name, str), "`bucket_name` must be a data bucket name string"

session = boto3.Session(region_name=region)
s3 = session.resource("s3")

In [12]:
# Upload the file to S3:
s3.Bucket(bucket_name).Object(interactions_path).upload_file(interactions_path)
interactions_s3uri = f"s3://{bucket_name}/{interactions_path}"
%store interactions_s3uri
print(f"Uploaded interactions to {interactions_s3uri}")

Stored 'interactions_s3uri' (str)
Uploaded interactions to s3://090247010259-ap-southeast-1-personalizepocvod1/poc_data/interactions.csv


Great - now our only *mandatory* dataset is ready to go! But, to try and improve our model, let's see if we can also prepare an item metadata file from the 'movies.csv' provided in the source dataset:

## Preparing (Item) Metadata

As before, we'll start out by loading and exploring the source data file:

In [13]:
original_items = pd.read_csv(dataset_dir + "/movies.csv")

print(original_items.info())
original_items.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB
None


movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [14]:
original_items.describe()

movieId
count    9742.000000
mean    42200.353623
std     52160.494854
min         1.000000
25%      3248.250000
50%      7300.000000
75%     76232.000000
max    193609.000000

This time, `movieId` is the only numeric column and so the only one we can generate summary statistics for. we also see `title` (A string title which appears to include the year of the film in brackets at the end) and `genres` - a bar-separated list of multiple genre tags for each film.

Amazon Personalize is already able to consume multi-categorical data in this bar-separated format, so we can use this `GENRES` field as-is.

Let's:

- Rename the `movieId` column to `ITEM_ID` in line with Amazon Personalize's required schema
- Rename our other columns to uppercase for consistency
- Try using a [regular expression](https://docs.python.org/3/library/re.html#module-re) to **extract the year** from the title field

with the code below:

In [15]:
items_df = original_items.copy()
items_df.rename(
    columns={
        "movieId": "ITEM_ID",
        "title": "TITLE",
        "genres": "GENRES",
    },
    inplace=True,
)
items_df["YEAR"] = items_df["TITLE"].str.extract(r".*(\d{4})")
items_df["YEAR"] = pd.to_numeric(items_df["YEAR"]).astype("Int64")

print(items_df.info())
items_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ITEM_ID  9742 non-null   int64 
 1   TITLE    9742 non-null   object
 2   GENRES   9742 non-null   object
 3   YEAR     9730 non-null   Int64 
dtypes: Int64(1), int64(1), object(2)
memory usage: 314.1+ KB
None


ITEM_ID                               TITLE  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        GENRES  YEAR  
0  Adventure|Animation|Children|Comedy|Fantasy  1995  
1                   Adventure|Children|Fantasy  1995  
2                               Comedy|Romance  1995  
3                         Comedy|Drama|Romance  1995  
4                                       Comedy  1995

This seems to have gone pretty well overall - but we can see from the summary statistics there are a small proportion of null values in our new `YEAR` column: Let's quickly check whether these gaps make sense:

In [16]:
items_df[items_df["YEAR"].isnull()].sample(5)

ITEM_ID              TITLE              GENRES  YEAR
9179   149334  Nocturnal Animals      Drama|Thriller  <NA>
9091   143410         Hyena Road  (no genres listed)  <NA>
6059    40697          Babylon 5              Sci-Fi  <NA>
9367   162414          Moonlight               Drama  <NA>
9259   156605           Paterson  (no genres listed)  <NA>

These examples seem to make sense as the relevant titles don't obviously list a year.

The proportion of missing values in the `YEAR` field is quite small, and the age of each movie is likely to be usefully descriptive - so we'll keep and use this field.

However, at this time the `TITLE` field is pretty useless to us from a modelling perspective, as today Amazon Personalize can only interpret string fields as category identifiers, not use natural language features mentioned within them.

...So as a final transformation we'll drop the `TITLE` column, and then our item metadata will be ready to load for modelling:

In [17]:
items_df.drop("TITLE", axis=1, inplace=True)
items_df.head()

ITEM_ID                                       GENRES  YEAR
0        1  Adventure|Animation|Children|Comedy|Fantasy  1995
1        2                   Adventure|Children|Fantasy  1995
2        3                               Comedy|Romance  1995
3        4                         Comedy|Drama|Romance  1995
4        5                                       Comedy  1995

As for the interactions data, we now just need to save this dataset to CSV format and upload to our Amazon S3 Bucket:

In [18]:
items_filename = "item-meta.csv"
items_path = f"{data_dir}/{items_filename}"
%store items_path

items_df.to_csv(items_path, index=False)

Stored 'items_path' (str)


In [19]:
# Upload the file to S3:
s3.Bucket(bucket_name).Object(items_path).upload_file(items_path)
items_s3uri = f"s3://{bucket_name}/{items_path}"
%store items_s3uri
print(f"Uploaded item metadata to {items_s3uri}")

Stored 'items_s3uri' (str)
Uploaded item metadata to s3://090247010259-ap-southeast-1-personalizepocvod1/poc_data/item-meta.csv


## All set!

We've now prepared interaction and item metadata in a compatible format for Amazon Personalize, and uploaded it to Amazon S3 ready for import.

In the next notebook we'll start actually using the Personalize service, and you have two choices:

- Follow along in the **AWS Console** with the instructions and screenshots in [02a_Importing_Data_(Console).ipynb](02a_Importing_Data_(Console).ipynb), *OR*
- Run the same steps in code with the **AWS SDK for Python (Boto3)** by following [02b_Importing_Data_(Python_SDK).ipynb](02b_Importing_Data_(Python_SDK).ipynb)